solutions/izho-17-LongestBeautifulSequence.mdx

```diff
@@ -0,0 +1,213 @@
+
+$a[i]$.
+
+$$
+\texttt{dp}[i]=\max_{j<i,\text{popcount}(a[i]\&a[j])=k[i]}\{\texttt{dp}[j]\}+1
+$$
+
+We can also use $a[i]$ as indices:
+
+$\mathcal{O}(N\max a[i])$ DP:
+
+Let $\texttt{dp}[i][j]$ be the length of longest beautiful sequence of the first $i$
+elements that ends with $j$,
+
+
+then
+$\max\limits_{j\leq\max{a[i]},\text{popcount}(j\&a[i])=k[i]}\{\texttt{dp}[i-1][j]\}+1$
+will be the length of longest beautiful sequence end with $a[i]$.
+
+$$
+\texttt{dp}[i][a[i]] = \max \{ \texttt{dp}[i-1][a[i]] , \max_{j\leq\max{a[i]},\text{popcount}(j\&a[i])=k[i]}\{\texttt{dp}[i-1][j]\}+1 \}
+$$
+
+$$
+\texttt{dp}[i][j] = \texttt{dp}[i-1][j] , j\not=a[i]
+$$
+
+Or another way:
+
+Let $\texttt{dp}[i][j][q]$ be the length of longest beautiful sequence of the first
+$i$ elements which satisfies $\text{popcount}(\texttt{EndElement}\&j)=q$,
+
+
+then $\texttt{dp}[i-1][a[i]][k[i]]+1$ will be the length of longest beautiful
+sequence end with $a[i]$.
+
+$$
+\texttt{dp}[i][j][\text{popcount}(a[i]\&j)]=\max_{j\leq\max{a[i]}}\{\texttt{dp}[i-1][j][\text{popcount}(a[i]\&j)],\texttt{dp}[i-1][a[i]][k[i]]+1\}
+$$
+
+$$
+\texttt{dp}[i][j][q]=\texttt{dp}[i-1][j][q],q\not=\text{popcount}(a[i]\&j)
+$$
+
+According to Hint 2, we need to consider either of them, that is, to consider
+the answers before and after.
+
+Since $a[i] < 2^{20}$ , divide the 20-bits number into two 10-bits numbers.
+
+For convenience, let $\text{b0}(x) = \lfloor \frac{x}{2^{10}} \rfloor$
+represents 10-19 bits, let $\text{b1}(x) = x \& 2^{10} $ represents 0-9 bits.
+
+
+Let $\texttt{ans}[i]$ be the length of longest beautiful sequence that ends with
+$a[i]$, and $\texttt{dp}[p][i][j][q]$ be the length of longest beautiful
+sequence of the first $p$ elements which satisfies $i = \text{b1}(\texttt{EndElement})$,
+
+$
+\text{popcount}(\text{b0}(\texttt{EndElement})\&j)=q$.
+
+For convenience, $\texttt{cur0} = \text{b0}(a[i])$,
+
+$\texttt{cur1} = \text{b1}(a[i]))$.
+
+Thus, we have
+
+
+$$
+\texttt{ans}[i] = \max_{\texttt{pre1} < 2^{10}} \{ \texttt{dp}[i-1][\texttt{pre1}][\texttt{cur0}][k[i] - \text{popcount}(\texttt{pre1} \& \texttt{cur1})] + 1 \}
+$$
+
+$$
+\texttt{dp}[i][\texttt{cur1}][\texttt{nex0}][\text{popcount}(\texttt{nex0} \& \texttt{cur0})] = \max \{ \texttt{dp}[i-1][\texttt{cur1}][\texttt{nex0}][\text{popcount}(\texttt{nex0} \& \texttt{cur0})] , \texttt{ans}[i] \}
+$$
+
+and
+
+$$
+\texttt{dp}[p][i][j][q] = \texttt{dp}[p-1][i][j][q]
+$$
+
+for $\texttt{dp}[p][i][j][q]$ which is not updated above.
+
+For the solution, just store which state this state is transferred from, then
+
+output in reverse order.
+
+</Spoiler>
+
+<Spoiler title="Implementation">
+
+**Time Complexity:** $\mathcal{O}(N\sqrt{\max{a[i]}})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+const int maxcount = 10;
+const int mask = 1 << maxcount;
+const int maxn = 100005;
+
+int cnt[mask];  // popcount array

```

Comment:
From: https://usaco.guide/general/adding-solution?lang=cpp

![image](https://user-images.githubusercontent.com/79723780/169675481-d063c0e0-718d-4d53-a398-d0f0d6431d3e.png)

Reply:
but it's constant